File: medium/Hash-Table/isomorphic-strings/is_isomorphic.cpp
#include <string>
#include <unordered_map>
using namespace std;

class Solution {
public:
    bool isIsomorphic(string s, string t) {
        /*
         * Check if two strings are isomorphic.
         * Approach: Two hash maps for bidirectional character mapping
         * Time Complexity: O(n)
         * Space Complexity: O(1)
         */
        unordered_map<char, char> s2t;
        unordered_map<char, char> t2s;

        for (int i = 0; i < s.size(); ++i) {
            char c1 = s[i];
            char c2 = t[i];

            if (s2t.count(c1) && s2t[c1] != c2)
                return false;
            if (t2s.count(c2) && t2s[c2] != c1)
                return false;

            s2t[c1] = c2;
            t2s[c2] = c1;
        }

        return true;
    }
};

File: medium/Hash-Table/isomorphic-strings/is_isomorphic.py
class Solution:
    def isIsomorphic(self, s: str, t: str) -> bool:
        """
        Determine if two strings are isomorphic.
        Approach: Two hash maps for bidirectional mapping
        Time Complexity: O(n)
        Space Complexity: O(1)
        """
        map_s_t = {}
        map_t_s = {}

        for c1, c2 in zip(s, t):
            # Check mapping from s to t
            if c1 in map_s_t and map_s_t[c1] != c2:
                return False
            # Check mapping from t to s
            if c2 in map_t_s and map_t_s[c2] != c1:
                return False

            map_s_t[c1] = c2
            map_t_s[c2] = c1

        return True
